src/core/allow-bind.c

```diff
@@ -0,0 +1,188 @@
+        if (map_fd < 0)
+                return log_unit_error_errno(u, map_fd, "Failed to get BPF map fd");
+
+        SET_FOREACH(port, ports_v4)
+                if (bpf_map_update_elem(map_fd, &port, &dummy, BPF_ANY))
+                        return log_unit_error_errno(u, -1, "Failed to update BPF map");
+
+        *ret_obj = TAKE_PTR(obj);
+        return 0;
+}
+
+int allow_bind_supported(void) {
+        _cleanup_(allow_bind_bpf_freep) struct allow_bind_bpf *obj = NULL;
+
+        int r = cg_unified_controller(SYSTEMD_CGROUP_CONTROLLER);
+        if (r < 0)
+                return log_error_errno(r, "Can't determine whether the unified hierarchy is used: %m");
+
+        if (r == 0) {
+                log_debug_errno(SYNTHETIC_ERRNO(EOPNOTSUPP),
+                                "Not running with unified cgroup hierarchy, BPF is not supported");
+                return 0;
+        }
+
+        if (!bpf_probe_prog_type(BPF_PROG_TYPE_CGROUP_SOCK_ADDR, /*ifindex=*/0)) {
+                log_debug_errno(SYNTHETIC_ERRNO(EOPNOTSUPP),
+                                "BPF program type cgroup_sock_addr is not supported");
+                return 0;
+        }
+
+        r = prepare_allow_bind_bpf(NULL, NULL, NULL, &obj);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to load BPF object: %m");
+
+        return can_link_bpf_program(obj->progs.allow_bind_v6);
+}
+
+static int allow_bind_install_impl(Unit *u) {
+        _cleanup_(bpf_link_freep) struct bpf_link *ipv4 = NULL, *ipv6 = NULL;
+        _cleanup_(allow_bind_bpf_freep) struct allow_bind_bpf *obj = NULL;
+        _cleanup_free_ char *cgroup_path = NULL;
+        _cleanup_close_ int cgroup_fd = -1;
+        CGroupContext *cc;
+        int r;
+
+        cc = unit_get_cgroup_context(u);
+        if (!cc)
+                return 0;
+
+        r = cg_get_path(SYSTEMD_CGROUP_CONTROLLER, u->cgroup_path, NULL, &cgroup_path);
+        if (r < 0)
+                return log_unit_error_errno(u, r, "Failed to get cgroup path: %m");
+
+        if (!cc->allow_bind_ports)
+                return 0;
+
+        r = prepare_allow_bind_bpf(u, cc->allow_bind_ports, cc->allow_bind_ports, &obj);
+        if (r < 0)
+                return log_unit_error_errno(u, r, "Failed to load BPF object: %m");
+
+        cgroup_fd = open(cgroup_path, O_RDONLY | O_CLOEXEC, 0);
+        if (cgroup_fd < 0)
+                return -errno;
+
+        ipv4 = bpf_program__attach_cgroup(obj->progs.allow_bind_v4, cgroup_fd);
+        if (libbpf_get_error(ipv4))
+                return log_unit_error_errno(u, r, "Failed to create ipv4 cgroup link");
+
+        ipv6 = bpf_program__attach_cgroup(obj->progs.allow_bind_v6, cgroup_fd);
+        if (libbpf_get_error(ipv6))
+                return log_unit_error_errno(u, r, "Failed to create ipv6 cgroup link");
+
+        u->ipv4_allow_bind_bpf_link = TAKE_PTR(ipv4);
+        u->ipv6_allow_bind_bpf_link = TAKE_PTR(ipv6);
```

Comment:
I'm wondering if it's possible to use [`bpf_link__update_program()`](https://github.com/libbpf/libbpf/blob/adaf538bcac57273bbb151915f30642eb3e06831/src/libbpf.h#L230) to avoid creating a new link each time but just updating its program instead. 

Also, this would be beneficial for reload/reexec, we can serialize the fd of the link and just update the program of it on the new process, in that way we avoid creating/destroying new link as well as having two links attached at the same time (for a very period of time while the new program is loaded and attached). 

One possible implementation could be to only store the fd of the link and update [`bpf_link_update`](https://github.com/libbpf/libbpf/blob/adaf538bcac57273bbb151915f30642eb3e06831/src/bpf.h#L194) to attach to the new program. Do you think it's worth to try it out?

Reply:
@mauriciovasquezbernal @anakryiko

Good idea, it can be extended.

I'll add a helper to libbpf to construct `bpf_link` from fd. With that helper it would be possible to construct links after daemon-reload, daemon-reexec instead of [storing fds in unit](https://github.com/systemd/systemd/pull/17655/files#diff-d81d5336a25405c5d3aa6cdf13fb83ee69b1d770250e4aae5cb594f4f0867132R311). 
Then `bpf_link__update_program` will be super helpful to atomically update the links with fds of newly loaded programs.
IIUC, this helper would go in libbpf 0.4.0.

Since both `u->ipv4_allow_bind_bpf_link` and ` u->ipv6_allow_bind_bpf_link` are `NULL` until they are assigned in `install` function, this doesn't make much difference now.